### DISCUSSION TITLE
Get an iterator from rust to R

### DISCUSSION THREAD

**Discussion by daniellga** [extendr/extendr#387] [Q&A]
Hi! I would like to call a rust `Iter` object from R, kind of similar to what Reticulate [does](https://rstudio.github.io/reticulate/reference/iterate.html). While `iterator_next` works, I am having some trouble trying to export `iterator_next_r` to R since it complains about the lifetime when I use `#[extendr]` above it. Any suggestions to help me?

```
fn vec_to_iter<'a>(vector: &'a Vec<i32>) -> std::slice::Iter<'a, i32> {
    vector.iter()
}

fn iterator_next<'a>(iterator: &mut std::slice::Iter<'a, i32>) -> Option<&'a i32> {
    iterator.next()
}

// #[extendr]
// fn iterator_next_r<'a>(iterator: &mut std::slice::Iter<'a, i32>) -> Robj {
//     match iterator.next() {
//         Some(x) => Robj::try_from(&x).expect("error"),
//         None => r!(NA_INTEGER)
//     }
// }

#[test]
fn test() {
    let v = vec![1,2,3];

    let mut a = vec_to_iter(&v);

    assert_eq!(Some(&1), iterator_next(&mut a));
    assert_eq!(Some(&2), iterator_next(&mut a));
    assert_eq!(Some(&3), iterator_next(&mut a));
    assert_eq!(None, iterator_next(&mut a));
    assert_eq!(None, iterator_next(&mut a));
}
```

error:
```
use of undeclared lifetime name `'a`
undeclared lifetimerustc[E0261](https://doc.rust-lang.org/error-index.html#E0261)
[lib.rs(173, 19): ]()consider introducing lifetime `'a` here: `<'a>`
```

Thank you!

**Comment by andy-thomason** [comment]
You can return by value as long as the type is static. IE no lifetimes.

You can then call next on the object to get the next value.

Ideally an Option should return NULL or the value. I may need to do some work on return types. I'll try to generate an example on Monday...

**Reply by andy-thomason** [reply]
In this case you need to own the vector or use an RC or Arc to make it static.

**Comment by daniellga** [comment]
Thank you. I will try to do as you suggested.

**Comment by andy-thomason** [comment]
I'm making a project to wrap rust-bio to find out what we need in terms of conversions
and other things.

I've discovered a few things to fix along the way.
